### PR TITLE
SI-6142: warn @inline-methods ending up not inlined (rightfully or not)

### PR DESCRIPTION
**Pull request by magarciaEPFL**
review by @paulp , @VladUreche

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/icode/analysis/TypeFlowAnalysis.scala [hunk 1]
```diff
@@ -470,6 +470,8 @@ abstract class TypeFlowAnalysis {
 
     val isOnWatchlist = mutable.Set.empty[Instruction]
 
+    val warnIfInlineFails = mutable.Set.empty[opcodes.CALL_METHOD] // cache for a given IMethod (ie cleared on Inliner.analyzeMethod).
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/backend/icode/analysis/TypeFlowAnalysis.scala:473]: Can we have a more descriptive name for this, maybe mustInlineCalls?

**Comment by VladUreche** [comment]
Miguel, does this mean we'll have a double warning for @inline-marked methods?
One given by warnNoInline and another one given by the new warning? Or am I missing something?

File: src/compiler/scala/tools/nsc/backend/icode/analysis/TypeFlowAnalysis.scala (hunk 1 above)
Comment (paulp) [inline src/compiler/scala/tools/nsc/backend/icode/analysis/TypeFlowAnalysis.scala:473]: Also, in my opinion it's good practice to use perRunCaches whenever possible.

**Comment by paulp** [comment]
Assuming the effect is what we're after, it looks good to me.

**Comment by magarciaEPFL** [comment]
@VladUreche , Before this commit, reporting of "un-inlineability" involves interpreting the result of `Inliner.lookupIMethod()` and `CallerCalleeInfo.isStampedForInlining()` , both at `analyzeInc()` time.

Those callsites which are discarded before `analyzeInc()` could inspect them resulted in no warning (e.g., some callsites are discarded, for efficiency reasons, as early as `putOnRadar()`).

As you mention, it's possible with this commit for two warnings to be emitted for a not-inlined callsite, however each provides different information: one includes the reason (extracted from `Inliner.lookupIMethod()` and `CallerCalleeInfo.isStampedForInlining()`) and the other summarizes in an unmistaken manner that an `@inline` target was not inlined, as requested by SI-6142.

**Comment by VladUreche** [comment]
In that case, LGTM

**Pull request merged**